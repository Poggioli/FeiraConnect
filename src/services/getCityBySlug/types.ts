import { UndefinedInitialDataOptions } from "@tanstack/react-query";
import { AxiosError } from "axios";
import { GetCities } from "@/services/getCities";

export type UseGetCityBySlug = GetCities;

export type UseGetCityBySlugQueryKey = ["getCityBySlug", string]

export type UseGetCityBySlugOptions = UndefinedInitialDataOptions<
  UseGetCityBySlug,
  AxiosError,
  UseGetCityBySlug,
  UseGetCityBySlugQueryKey
>;
