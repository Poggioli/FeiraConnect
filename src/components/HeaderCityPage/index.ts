export * from "./HeaderCityPage";
export * from "./useHeaderCityPage";