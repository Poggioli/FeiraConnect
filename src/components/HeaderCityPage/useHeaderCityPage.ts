import {
  createGetCityBySlugQueryKey,
  useGetCityBySlug,
} from "@/services/getCityBySlug";
import { useParams } from "@tanstack/react-router";

export function useHeaderCityPage() {
  const { city } = useParams({ from: "/city/$city" });
  const cityService = useGetCityBySlug({
    queryKey: createGetCityBySlugQueryKey(city),
  });

  return {
    cityService,
  };
}
