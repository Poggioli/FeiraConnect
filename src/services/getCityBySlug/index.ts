export * from "./getCityBySlug";
export * from "./types";