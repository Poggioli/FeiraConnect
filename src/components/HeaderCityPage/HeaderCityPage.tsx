import { TitlePage } from "@/components/TitlePage";
import { Skeleton } from "@/components/ui/skeleton";
import { notFound } from "@tanstack/react-router";
import { FC } from "react";
import { useHeaderCityPage } from "./useHeaderCityPage";

export const HeaderCityPage: FC = () => {

  const {
    cityService: {
      isFetching,
      error,
      data,
    }
  } = useHeaderCityPage();

  if (error && error.response?.status === 404) {
    throw notFound({ routeId: '/city/$city' })
  }

  return isFetching ? (
    <Skeleton className="min-h-10 w-32" />
  ) : (
    <TitlePage>{data?.name}</TitlePage>
  );
};
