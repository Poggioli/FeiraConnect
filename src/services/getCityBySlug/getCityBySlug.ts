import { api } from "@/services/api";
import { useQuery } from "@tanstack/react-query";
import { AxiosError } from "axios";
import {
  UseGetCityBySlug,
  UseGetCityBySlugOptions,
  UseGetCityBySlugQueryKey,
} from "./types";

const queryKey = "getCityBySlug" as const;
const url = "cities" as const;

export function createGetCityBySlugQueryKey(
  slug: string
): UseGetCityBySlugQueryKey {
  return [queryKey, slug];
}

export function useGetCityBySlug(options: UseGetCityBySlugOptions) {
  return useQuery<
    UseGetCityBySlug,
    AxiosError,
    UseGetCityBySlug,
    UseGetCityBySlugQueryKey
  >({
    ...options,
    queryKey: options.queryKey,
    queryFn: async ({ queryKey: { "1": slug } }) => {
      const res = await api.get<UseGetCityBySlug>(`${url}/${slug}`);
      return res.data;
    },
  });
}
