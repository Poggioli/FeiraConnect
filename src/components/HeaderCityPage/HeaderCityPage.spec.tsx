import { waitFor } from "@testing-library/dom";
import { customRender } from "@tests/customRender";
import { getCityBySlugHandlers } from "@tests/getCityBySlug.mock";
import { server } from "@tests/server.mock";
import { HeaderCityPage } from "./HeaderCityPage";

const mockUseParams = vi.fn();
const mockNotFound = vi.fn();

vi.mock("@tanstack/react-router", async (importOriginal) => {
  const original = await importOriginal<Record<string, unknown>>();
  return {
    ...original,
    useParams: (args: unknown) => mockUseParams(args),
    notFound: (args: unknown) => mockNotFound(args)
  };
});

describe("@/components/HeaderCityPage", () => {
  beforeEach(() => {
    mockUseParams.mockReturnValue({ city: "slug" });
  });

  it(`GIVEN a HeaderCityPage
      WHEN server return success
      THEN should show the name of city`, async () => {
    server.use(getCityBySlugHandlers.filled);
    const { getByText } = customRender(<HeaderCityPage />);

    await waitFor(() => {
      const title = getByText("Slug 1");
      expect(title).toBeVisible();
    })
  });

  it(`GIVEN a HeaderCityPage
      WHEN not finish load
      THEN should show the loading state`, () => {
    server.use(getCityBySlugHandlers.filledWithDelay);
    const { baseElement } = customRender(<HeaderCityPage />);

    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a HeaderCityPage
      WHEN response is 404
      THEN should call notFound method from @tanstack/react-router`, async () => {
    server.use(getCityBySlugHandlers.error);
    const { getByText } = customRender(<HeaderCityPage />);

    await waitFor(() => {
      const errorMessage = getByText('Error Boundary')
      expect(errorMessage).toBeVisible();
      expect(mockNotFound).toHaveBeenCalled();
      expect(mockNotFound).toHaveBeenCalledWith({ routeId: "/city/$city" });
    })
  });
});
