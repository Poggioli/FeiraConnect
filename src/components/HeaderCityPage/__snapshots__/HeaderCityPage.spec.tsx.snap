// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@/components/HeaderCityPage > GIVEN a HeaderCityPage
      WHEN not finish load
      THEN should show the loading state 1`] = `
<body>
  <div />
</body>
`;
